### PR TITLE
Use sieve for prime(), composite(), isprime() etc. if sieve is large enough

### PR DESCRIPTION
**Pull request by postvakje**
If sieve is large enough, use it to compute the functions prime(),
composite() , primepi(), compositepi(). prevprime(), nextprime() and
isprime().

### CODE REVIEW

**Comment by asmeurer** [comment]
It's not obvious to me that composite and isprime are faster. Then again, the sieve will generally be relatively small, so it probably won't make a huge difference anyway.

**Comment by postvakje** [comment]
The benefits are probably apparent when the sieve (which was computed elsewhere prior) is very  large.  With a precomputed sieve, the time for sieve-based `isprime` is log(n/log n) versus k(log n)^3 for MR.  If one needs to call `isprime` many times over the same set of numbers, it might be more efficient to extend the sieve once and then call `isprime`.

File: sympy/ntheory/generate.py [hunk 1]
```diff
@@ -247,6 +247,8 @@ def prime(nth):
     prime_arr = [2, 3, 5, 7, 11, 13, 17]
     if n <= 7:
         return prime_arr[n - 1]
+    if n <= len(sieve._list):
```
Comment (smichr) [inline sympy/ntheory/generate.py]: Should the use of `prime_arr` be removed in favor of the sieve?

**Comment by smichr** [comment]
I improved coverage and made other minor changes [here](https://github.com/postvakje/sympy/pull/3).

**Comment by postvakje** [comment]
@smichr I didn't realize you have already removed prime_arr before I sent another commit :-)

**Comment by smichr** [comment]
No problem. I will commit as is since I don't think @asmeurer comment is a blocker.

**Pull request merged**